zilliqa/src/bin/validator_oracle.rs

```diff
@@ -0,0 +1,238 @@
+use std::{path::PathBuf, str::FromStr};
+
+use alloy_provider::{Provider, ProviderBuilder, WsConnect};
+use alloy_rpc_types::{BlockNumberOrTag, Filter};
+use anyhow::Result;
+use clap::Parser;
+use ethers::{
+    contract::abigen,
+    providers::Middleware,
+    signers::{LocalWallet, Signer},
+    types::{TransactionRequest, H160},
+};
+use futures_util::stream::StreamExt;
+use tokio::sync::watch;
+use tracing::{debug, error, info};
+use tracing_subscriber::EnvFilter;
+use zilliqa::{
+    contracts,
+    crypto::{NodePublicKey, SecretKey},
+    state::contract_addr,
+    uccb::{
+        cfg::{ChainConfig, Config, ZQ2Config},
+        client::ChainClient,
+    },
+};
+
+// The json files are generated by invoking 'forge build' in zilliqa/src/contracts/uccb
+abigen!(
+    ValidatorManager,
+    "zilliqa/src/contracts/uccb/out/ValidatorManager.sol/ValidatorManager.json"
+);
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(value_parser = SecretKey::from_hex)]
+    secret_key: SecretKey,
+    #[clap(long, short, default_value = "config.toml")]
+    config_file: PathBuf,
+}
+
+impl zilliqa::uccb::Args for Args {
+    fn secret_key(&self) -> &SecretKey {
+        &self.secret_key
+    }
+
+    fn config_file(&self) -> &PathBuf {
+        &self.config_file
+    }
+}
+
+struct ValidatorOracle {
+    wallet: LocalWallet,
+    zq2_config: ZQ2Config,
+    validators: Vec<NodePublicKey>,
+    chain_clients: Vec<ChainClient>,
+}
+
+impl ValidatorOracle {
+    pub async fn new(secret_key: SecretKey, config: Config) -> Result<Self> {
+        let wallet = LocalWallet::from_str(secret_key.to_hex().as_str())?;
+        let mut chain_clients = vec![Self::create_zq2_chain_client(config.clone(), &wallet).await?];
+        for chain_config in config.chain_configs {
+            chain_clients.push(
+                ChainClient::new(
+                    &chain_config,
+                    H160(**config.zq2.validator_manager_address),
+                    wallet.clone(),
+                )
+                .await?,
+            );
+        }
+
+        Ok(Self {
+            wallet,
+            zq2_config: config.zq2,
+            validators: vec![],
+            chain_clients,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<()> {
+        info!(
+            "Starting validator oracle with wallet address {}",
+            self.wallet.address()
+        );
+
+        self.validators = self.get_stakers().await?;
+        info!("Validators: {}\n", serde_json::to_string(&self.validators)?);
```

Comment:
Log messages shouldn't have an explicit newline and we shouldn't need to rely on `serde_json::to_string`. Can this be something like `info!(?self.validators)` ?